src/main/java/baritone/cache/FasterWorldScanner.java

```diff
@@ -0,0 +1,271 @@
+                    queued++;
+                    cachedWorld.queueForPacking(chunk);
+                }
+            }
+        }
+
+        return queued;
+    }
+
+    // ordered in a way that the closest blocks are generally first
+    public static List<ChunkPos> getChunkRange(int centerX, int centerZ, int chunkRadius) {
+        List<ChunkPos> chunks = new ArrayList<>();
+        // spiral out
+        chunks.add(new ChunkPos(centerX, centerZ));
+        for (int i = 1; i < chunkRadius; i++) {
+            for (int j = 0; j <= i; j++) {
+                chunks.add(new ChunkPos(centerX - j, centerZ - i));
+                if (j != 0) {
+                    chunks.add(new ChunkPos(centerX + j, centerZ - i));
+                    chunks.add(new ChunkPos(centerX - j, centerZ + i));
+                }
+                chunks.add(new ChunkPos(centerX + j, centerZ + i));
+                if (j != i) {
+                    chunks.add(new ChunkPos(centerX - i, centerZ - j));
+                    chunks.add(new ChunkPos(centerX + i, centerZ - j));
+                    if (j != 0) {
+                        chunks.add(new ChunkPos(centerX - i, centerZ + j));
+                        chunks.add(new ChunkPos(centerX + i, centerZ + j));
+                    }
+                }
+            }
+        }
+        return chunks;
+    }
+
+    private List<BlockPos> scanChunksInternal(IPlayerContext ctx, BlockOptionalMetaLookup lookup, List<ChunkPos> chunkPositions, int maxBlocks) {
+        assert ctx.world() != null;
+        try {
+            Stream<BlockPos> posStream = chunkPositions.parallelStream().flatMap(p -> scanChunkInternal(ctx, lookup, p));
+            if (maxBlocks >= 0) {
+                // WARNING: this can be expensive if maxBlocks is large...
+                // see limit's javadoc
+                posStream = posStream.limit(maxBlocks);
+            }
+            return posStream.collect(Collectors.toList());
+        } catch (Exception e) {
+            e.printStackTrace();
+            throw e;
+        }
+    }
+
+    private Stream<BlockPos> scanChunkInternal(IPlayerContext ctx, BlockOptionalMetaLookup lookup, ChunkPos pos) {
+        IChunkProvider chunkProvider = ctx.world().getChunkProvider();
+        // if chunk is not loaded, return empty stream
+        if (!chunkProvider.isChunkGeneratedAt(pos.x, pos.z)) {
+            return Stream.empty();
+        }
+
+        long chunkX = (long) pos.x << 4;
+        long chunkZ = (long) pos.z << 4;
+
+        int playerSectionY = ctx.playerFeet().y >> 4;
+
+        List<BlockPos> blocks = new ArrayList<>();
+
+        collectChunkSections(lookup, chunkProvider.getLoadedChunk(pos.x, pos.z), playerSectionY, (section, isInFilter) -> {
+            int yOffset = section.getYLocation();
+            BitArray array = (BitArray) ((IBlockStateContainer) section.getData()).getStorage();
+            collectBlockLocations(array, isInFilter, place -> blocks.add(new BlockPos(
```

Comment:
This lambda here `... place -> blocks.add ...` is impure - it captures the context of `blocks` and other variables. For this reason, it'll allocate the lambda each time on the heap. If the render distance was large (say, 32 chunks in each direction), and there were 16 sections in each chunk (24 in newer versions of mc), then that's 98304 allocations. if we are doing something like mineprocess where we async scan https://github.com/cabaletta/baritone/blob/bc8c823045e380c2d77f4361a7e0d86c6db476e6/src/api/java/baritone/api/Settings.java#L834 four times a second, that's **393,216 objects** allocated on the heap per second. This is enough to strain GC, on my laptop at least.

I think at least we'd want to pass through these parameters (`blocks', `chunkX`, `yOffset`, `chunkZ`) through directly. But really that would be gross. I think this pattern of passing around a consumer is not worth it. In my opinion it's not even easier to understand or more elegant, rather the opposite :(